### PR TITLE
feat!: add typed query context parameters

### PR DESCRIPTION
**Pull request by FrankChen021**
### Description

This PR introduces typed, centrally declared query context parameters and migrates two representative parameters end to end:

- `useResultLevelCache`, a Boolean parameter with a declared default.
- `maxRowsQueuedForOrdering`, an Integer parameter with a runtime fallback and a range constraint.

The goal is to establish and demonstrate the parameter model with a small, reviewable change before migrating the remaining query context parameters.

### Roadmap
After this PR is accepted, we will move other existing query context paramters into the centralized class in serveral PRs and finally we will provide a `sys.query_context_parameters` upon this centralized configuration.

### Background

[Issue #17769](https://github.com/apache/druid/issues/17769) proposes improving query context discoverability and validation. The closed [PR #18087](https://github.com/apache/druid/pull/18087) explored a centralized registry, a system table, and broad migration of existing parameters, but touched more than 100 files.

This PR extracts a smaller end-to-end foundation from that work. It deliberately catalogs only two parameters with different types and migrates their call sites completely.

**Complete validation of SQL `SET` parameters is not implemented in this PR.** The `SET` processing path calls the catalog validation API to demonstrate the intended integration, but only the two migrated parameters are recognized. Unknown and unmigrated parameter names continue to be accepted until the catalog migration is complete.

#### Parameter model and catalog

Adds `QueryContextParameter<T>` and the centralized `QueryContextParameters` catalog. A descriptor contains its name, Java type, parser, nullability, optional default, constraints, deprecation message, and documentation metadata. The immutable `BY_NAME` catalog is derived from the declared public parameter fields.

#### Typed context APIs and call-site migration

Migrated these two parameters to demonstrate the APIs of above models:

- `useResultLevelCache`, a Boolean parameter with a declared default.
- `maxRowsQueuedForOrdering`, an Integer parameter with a runtime fallback and a range constraint.

#### Documentation generation

Adds a compile-time generator for descriptor-backed rows in the query context and Scan query documentation.

- Normal Maven builds use `verify` mode. The generator renders complete document copies under `processing/target/generated-docs`, compares them with the checked-in Markdown, and fails the build if a generated row is stale. It does not modify source documentation in this mode.
- Running Maven with `-Dquery.context.docs.mode=generate` updates the checked-in Markdown. The generator replaces only lines ending in an exact `GENERATED QUERY CONTEXT PARAMETER` marker. This currently covers the `useResultLevelCache` row in `query-context-reference.md` and the `maxRowsQueuedForOrdering` row in `scan-query.md`.
- All other document lines are retained. Because generate mode rewrites each affected source file, it also normalizes line endings to `\n`.

#### SQL `SET` integration hook

Connects `SqlQueryPlus` to catalog validation. This validates recognized, migrated parameters and intentionally accepts all others. It demonstrates the future validation flow without claiming complete `SET` validation.

### CODE REVIEW

**Review by FrankChen021 — COMMENTED** [review]
Review complete: no high-confidence correctness, security, or reliability issues found in the current changes.

Reviewed 30 of 30 changed files.

Validation: focused `git diff --no-ext-diff --check` passed. Builds and tests were not run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java [hunk 1]
```diff
@@ -272,6 +273,7 @@ public Builder auth(final AuthenticationResult authResult)
     public SqlQueryPlus build()
     {
       final StatementAndSetContext statementAndSetContext = DruidSqlParser.parse(sql, true);
+      QueryContextParameters.validate(statementAndSetContext.getSetContext());
```
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java]: [P1] Preserve bad-request semantics for invalid SET values

`QueryContextParameters.validate` can throw an ordinary `IAE` (for example, `SET maxRowsQueuedForOrdering = 0; SELECT ...`), and this call runs while `SqlResource.doPost` is still creating the `HttpStatement`. That handler treats every pre-statement exception that is not already a `DruidException` as an operator `RUNTIME_FAILURE`, so the new validation path returns HTTP 500 for malformed user input instead of the 400 invalid-SQL response used by the parser. Convert validation failures to `InvalidSqlInput.exception(...)` (or another `DruidException` with `INVALID_INPUT`) before they cross this HTTP boundary.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

The review found one high-confidence user-facing error-handling issue; see the inline finding above.

Reviewed 30 of 30 changed files.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java [hunk 2]
```diff
@@ -272,6 +274,12 @@ public Builder auth(final AuthenticationResult authResult)
     public SqlQueryPlus build()
     {
       final StatementAndSetContext statementAndSetContext = DruidSqlParser.parse(sql, true);
+      try {
+        QueryContextParameters.validate(statementAndSetContext.getSetContext());
+      }
+      catch (IllegalArgumentException e) {
```
Comment (FrankChen021) [inline sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java]: [P1] Preserve bad-request semantics for parser failures

QueryContextParameters.validate delegates the migrated integer and boolean descriptors to QueryContexts.getAsInt and getAsBoolean, which throw BadQueryContextException for malformed strings or wrong types. That exception is a QueryException, not an IllegalArgumentException, so this catch does not convert it to InvalidSqlInput. Since validation runs before HttpStatement creation, SqlResource.handleExceptionBeforeStatementCreated treats it as a non-Druid exception and returns an operator RUNTIME_FAILURE (HTTP 500) for inputs such as SET maxRowsQueuedForOrdering = 'not-an-int'; SELECT 1 or SET useResultLevelCache = 1; SELECT 1. Catch and convert parser failures here so all invalid recognized SET values remain HTTP 400 invalid SQL.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

The updated validation path still lets malformed recognized SET values become HTTP 500; see the inline finding.

Reviewed 31 of 31 changed files.



---

This is an automated review by Codex GPT-5.6-Luna(max)

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, security, compatibility, serialization, integration, and test coverage; no high-confidence issues found in the current changes.

Reviewed 31 of 31 changed files.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: docs/querying/query-context-reference.md [hunk 3]
```diff
@@ -51,7 +51,7 @@ Unless otherwise noted, the following parameters apply to all query types, and t
...
 |`useCache`         | `true`                                 | Flag indicating whether to leverage the query cache for this query. When set to false, it disables reading from the query cache for this query. When set to true, Apache Druid uses `druid.broker.cache.useCache` or `druid.historical.cache.useCache` to determine whether or not to read from the query cache |
 |`populateCache`    | `true`                                 | Flag indicating whether to save the results of the query to the query cache. Primarily used for debugging. When set to false, it disables saving the results of this query to the query cache. When set to true, Druid uses `druid.broker.cache.populateCache` or `druid.historical.cache.populateCache` to determine whether or not to save the results of this query to the query cache |
-|`useResultLevelCache`| `true`                      | Flag indicating whether to leverage the result level cache for this query. When set to false, it disables reading from the query cache for this query. When set to true, Druid uses `druid.broker.cache.useResultLevelCache` to determine whether or not to read from the result-level query cache |
```
Comment (GWphua) [inline docs/querying/query-context-reference.md:54]: Is this comment an intended addition?

File: docs/querying/scan-query.md [hunk 4]
```diff
@@ -197,7 +197,7 @@ Configuration properties:
 
 |property|description|values|default|
 |--------|-----------|------|-------|
-|maxRowsQueuedForOrdering|The maximum number of rows returned when time ordering is used.  Overrides the identically named config.|An integer in [1, 2147483647]|`druid.query.scan.maxRowsQueuedForOrdering`|
+|maxRowsQueuedForOrdering|The maximum number of rows returned when time ordering is used. Overrides the identically named config.|An integer in [1, 2147483647]|`druid.query.scan.maxRowsQueuedForOrdering`| <!-- GENERATED QUERY CONTEXT PARAMETER: maxRowsQueuedForOrdering -->
```
Comment (GWphua) [inline docs/querying/scan-query.md:200]: Same as above

File: processing/src/main/java/org/apache/druid/query/QueryContext.java [hunk 5]
```diff
@@ -81,11 +84,152 @@ public static QueryContext empty()
...
+  )
+  {
+    return builder()
+        .put(parameter1, value1)
+        .put(parameter2, value2)
+        .put(parameter3, value3)
+        .put(parameter4, value4)
+        .toMap();
+  }
+
```
Comment (GWphua) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:232]: These methods should be unnecessary. We can simply use the following instead:

QueryContext context = QueryContext.builder()
    .put(...)
    .put(...)
    .build();

File: processing/src/main/java/org/apache/druid/query/QueryContext.java [hunk 6]
```diff
@@ -199,18 +389,6 @@ public long getLong(final String key, final long defaultValue)
...
-   * context value is not set.
-   *
-   * @throws BadQueryContextException for an invalid value
-   */
-  @SuppressWarnings("unused")
-  public Float getFloat(final String key)
-  {
-    return QueryContexts.getAsFloat(key, get(key));
-  }
-
```
Comment (GWphua) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:213]: This removal seem out of scope of this PR

File: processing/src/main/java/org/apache/druid/query/QueryContext.java [hunk 7]
```diff
@@ -514,15 +692,6 @@ public long getTimeout(long defaultTimeout)
...
 
-  @Nullable
-  public Duration getTimeoutDuration()
-  {
-    if (hasTimeout()) {
-      return Duration.ofMillis(getTimeout());
-    }
-    return null;
-  }
-
```
Comment (GWphua) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:525]: This removal seem out of scope of this PR

File: processing/src/main/java/org/apache/druid/query/QueryContext.java [hunk 8]
```diff
@@ -818,12 +987,4 @@ public RealtimeSegmentsMode getRealtimeSegmentsMode()
...
   }
 
-  /**
-   * @deprecated Use {@link #getRealtimeSegmentsMode()} instead.
-   */
-  @Deprecated
-  public boolean isRealtimeSegmentsOnly()
-  {
-    return getRealtimeSegmentsMode() == RealtimeSegmentsMode.EXCLUSIVE;
-  }
```
Comment (GWphua) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:828]: This removal seem out of scope of this PR

File: processing/src/main/java/org/apache/druid/query/Query.java [hunk 9]
```diff
@@ -132,7 +134,7 @@ default QueryContext context()
    * {@link QueryContext#getString(String)} <br/>
    * {@link QueryContext#getInt(String)} <br/>
    * {@link QueryContext#getLong(String)} <br/>
-   * {@link QueryContext#getFloat(String)} <br/>
+ * {@link QueryContext#getFloat(String, float)} <br/>
```
Comment (GWphua) [inline processing/src/main/java/org/apache/druid/query/Query.java]: Indentation + Should revert if we choose not to delete QueryContext#getFloat(String)

File: server/src/main/java/org/apache/druid/client/CachingClusteredClient.java [hunk 10]
```diff
@@ -297,24 +298,24 @@ private class SpecificQueryRunnable<T>
...
 
       if (populateCache) {
         // prevent down-stream nodes from caching results as well if we are populating the cache
-        contextBuilder.put(CacheConfig.POPULATE_CACHE, false);
-        contextBuilder.put(QueryContexts.BY_SEGMENT_KEY, true);
+        contextBuilder.putRaw(CacheConfig.POPULATE_CACHE, false);
+        contextBuilder.putRaw(QueryContexts.BY_SEGMENT_KEY, true);
       }
-      return contextBuilder.build();
+      return contextBuilder.toMap();
```
Comment (GWphua) [inline server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:318]: Is this change necessary?

**Review by GWphua — COMMENTED** [review]
PR seems to have a lot of unrelated changes. Lots of them can be re-used by methods currently in the codebase. Can do a re-check of them. 

Unnecessary public APIs
- [QueryContext.of(...) and ofMap(...)](https://github.com/apache/druid/blob/dca26037c1b06ff59912dfcc5402a715cdea616e/processing/src/main/java/org/apache/druid/query/QueryContext.java#L100-L230): eight overloads covering one through four parameter/value pairs. Replace all with the builder.
- [Query.withOverriddenContext(parameter, value)](https://github.com/apache/druid/blob/dca26037c1b06ff59912dfcc5402a715cdea616e/processing/src/main/java/org/apache/druid/query/Query.java#L189-L202): no production caller; only new tests use it.
- Druids.TimeseriesQueryBuilder.context(parameter, value): no production caller and arbitrarily added only to the timeseries builder.
- QueryContexts.override(context, parameter, value): completely unused and conflicts with the other APIs’ null semantics.
- QueryContextParameter.set: production usage is only an edited embedded test.
- QueryContextParameter.parseOrDefault: used only by its own unit tests.
- QueryContext.has(parameter): only needed internally by get; it does not need to be public.
- QueryContextParameter.toString: used only by its test.
- longParameter and stringParameter: no descriptors use them.

**Review by GWphua — CHANGES_REQUESTED** [review]

File: docs/querying/query-context-reference.md (hunk 3 above)
Comment (FrankChen021) [reply docs/querying/query-context-reference.md:54]: the document is updated by the documentation generator which removes extra SPACEs

File: processing/src/main/java/org/apache/druid/query/QueryContext.java (hunk 5 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/QueryContext.java:232]: These overloads are intentional rather than accidental duplication. They are the typed replacement for `ImmutableMap.of`/`Map.of` at migrated call sites: `ofMap` keeps the existing `Map<String, Object>` shape, while `of` returns `QueryContext`. The follow-up query-context migrations will use these APIs, so removing them now would force those changes back to raw keys or repetitive builder code. The builder remains available for larger or mixed contexts.

File: processing/src/main/java/org/apache/druid/query/QueryContext.java (hunk 6 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/QueryContext.java:213]: This removal is intentional. `getFloat(String)` has no in-repository callers, and the typed descriptor accessors are the API we want to expose going forward. Removing this unused overload keeps `QueryContext`'s public API focused; it is not required by the follow-up parameter migrations.

File: processing/src/main/java/org/apache/druid/query/QueryContext.java (hunk 7 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/QueryContext.java:525]: This is intentional API cleanup: `getTimeoutDuration()` has no in-repository callers, and callers can derive a duration from `getTimeout()` when needed. We want to remove unused legacy convenience methods from `QueryContext` rather than carry them into the typed API.

File: processing/src/main/java/org/apache/druid/query/QueryContext.java (hunk 8 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/QueryContext.java:828]: `isRealtimeSegmentsOnly()` is deprecated and has no in-repository callers. `getRealtimeSegmentsMode()` is the replacement, so this removal is intentional to keep the cleaned-up API focused. It is not needed by the follow-up migrations.

File: processing/src/main/java/org/apache/druid/query/Query.java (hunk 9 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/Query.java]: Fixed the Javadoc indentation in commit [`350ce20635`](https://github.com/FrankChen021/druid/commit/350ce20635498de5f08acbccff9a8843c18f6ed9). The `getFloat(String)` removal is intentional for the unused-API cleanup described above.

File: server/src/main/java/org/apache/druid/client/CachingClusteredClient.java (hunk 10 above)
Comment (FrankChen021) [reply server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:318]: Yes, this change is intentional. Context construction should migrate from `ImmutableMap`/`Map` builders to the typed `QueryContextBuilder`/`QueryContext.of*` APIs. For parameters that are not descriptors yet, `putRaw` is the compatibility path; continuing to use raw maps would require repetitive `QueryContextParameters.X.getName()` calls and would not demonstrate the new API. Once the remaining descriptors are migrated, those raw entries can move to typed `put` calls.

File: docs/querying/query-context-reference.md (hunk 3 above)
Comment (FrankChen021) [reply docs/querying/query-context-reference.md:54]: Yes, the marker is intentional. `<!-- GENERATED QUERY CONTEXT PARAMETER: useResultLevelCache -->` is an invisible HTML comment used as a stable anchor: the compile-time generator finds the marker and replaces the complete table row from descriptor metadata, while preserving the surrounding hand-written documentation. It is not rendered as user-facing content. The same marker is used for generated rows in the scan and SQL context documents.

File: docs/querying/scan-query.md (hunk 4 above)
Comment (FrankChen021) [reply docs/querying/scan-query.md:200]: Same explanation: the generated marker is intentional and is the generator's replacement anchor; it remains invisible in rendered documentation.

**Review by FrankChen021 — COMMENTED** [review]
Thanks for the review. The APIs called out as unnecessary are intentional parts of the migration surface: typed `of`/`ofMap` replace immutable-map construction, typed `withOverriddenContext`, the typed builder context overload, `QueryContextParameter.set`, and `QueryContext.has` support upcoming production migrations, and `QueryContexts.override`, `parseOrDefault`, and the parameter builder helpers are retained for the follow-up descriptor migrations. The unused legacy `QueryContext` convenience methods (`getFloat(String)`, `getTimeoutDuration()`, and deprecated `isRealtimeSegmentsOnly()`) are deliberately removed to keep the public API focused. The `CachingClusteredClient` change demonstrates the same migration for context construction; `putRaw` is retained only where a descriptor is not available yet. Generated HTML markers are intentional invisible anchors used by the compile-time documentation generator.

Reviewed 31 of 31 changed files.



---

This is an automated review by Codex GPT-5.6-luna

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, security, compatibility, serialization, integration, and test coverage; no high-confidence issues found in the current changes.

Reviewed 31 of 31 changed files.



---

This is an automated review by Codex GPT-5.6-Luna(max)
